Enable iteration of the DB

JimmyPettersson85 opened the pull request:
Hi!

Dont know if you're accepting pull requests but here goes.

Commit:
Implemented Each() to enable iteration of the DB

Commit:
Added tests for Each()

Commit:
Added example for Each()

colinmarc commented:
You bet I am! Sorry in advance if I'm super picky in review.

colinmarc commented on cdb_test.go:
```diff
@@ -27,6 +27,18 @@ var expectedRecords = [][][]byte{
 	{[]byte("not in the table"), nil},
 }
 
+var expectedKeyOrder = [][]byte{
```
Can we just grab this out of the table of records above?

JimmyPettersson85 replied:
Sure, if I switch the last two inserted values the order will be correct

colinmarc commented on README.md:
```diff
@@ -41,4 +41,10 @@ if err != nil {
 // => Practice
+
+// Iterate over the database
+db.Each(func(key, value []byte) error {
```
Instead of this, can we add something like the "stateful iterators" (the fourth option) described [here](http://ewencp.org/blog/golang-iterators/)?

That's the most similar to the stdlib (see [scanner](https://golang.org/pkg/bufio/#Scanner), [sql](https://golang.org/pkg/database/sql/#Rows), etc), and I like it a bit better.

Then the example would look like:

``` go
iter := db.Iter()
for iter.Next() {
   log.Printf("The key %s has a value of length %d\n", iter.Key(), iter.Value())
}

if iter.Err() != nil {
  log.Fatal(err)
}
```

JimmyPettersson85 replied:
I'll have a look at it after the weekend, shouldn't be any problem

JimmyPettersson85 replied:
Got some time to look at it now actually. Do you want the iterator type defined in its own file or in cdb.go?

colinmarc replied:
own file makes sense to me.

colinmarc commented on cdb_test.go:
```diff
@@ -44,6 +56,23 @@ func TestGet(t *testing.T) {
+
+	n := 0
+	db.Each(func(key, value []byte) error {
+		assert.Equal(t, string(expectedKeyOrder[n]), string(key))
```
this should just be `expectedRecords[n][0]` instead of using `expectedKeyOrder`.

Commit:
Implemented new iterator functionality

Removed method Each() since it was made obsolete by this change.

Commit:
Added tests and benchmark for the iterator

Commit:
Updated README

JimmyPettersson85 commented:
Pushed new commits with the Iterator type and removed the Each() method from CDB

colinmarc commented on iterator.go:
```diff
@@ -0,0 +1,63 @@
+package cdb
+
+// Iterator represents an iterator over a CDB database. It is used to sequentially
+// iterate over all kv-pairs in the database.
```
how about just "Iterator represents a sequential iterator over a CDB database."

colinmarc commented on iterator.go:
```diff
@@ -0,0 +1,63 @@
+// and has more records to iterate over.
+// If Next returns false it either encountered an error (can be read through Err())
+// or it has no more records to iterate over.
+func (iter *Iterator) Next() bool {
```
"Next reads the next key/value pair and advances the iterator one record. It returns false when the scan stops, either by reaching the end of the database or an error. After Next returns false, the Err method will return any error that occurred while iterating."

Copied from the stdlib one =)

JimmyPettersson85 replied:
Which one in the stdlib? It seems to talk about both the Next() and Scan() methods. I have not implemented a Scan() so maybe something along the lines of

"Next reads the next key/value pair and advances the iterator one record. It returns false when the scan stops, either by reaching the end of the database or an error. After **Next** returns false, the Err method will return any error that occurred during scanning."

colinmarc replied:
sorry, yeah, edited my comment. That's what I get for copy/pasting.

colinmarc commented on iterator.go:
```diff
@@ -0,0 +1,63 @@
+	return iter.pos < iter.endPos
+}
+
+// Key returns the current key.
```
We should add something here (and to `Value`) like "The byte slice returned is only valid until the next call to Next."

colinmarc replied:
Actually, that's not true, nevermind.

colinmarc commented:
This is looking great! Just some doc updates. Can you add a testable example to `cdb_test.go`, too? And then I think this is ready.

JimmyPettersson85 commented:
I have updated the docs according to above and refactored out the iterator test/benchmark into a new file iterator_test.go instead of having them in cdb_test.go.

Will this do as an example?

``` go
func ExampleIterator() {
    db, err := cdb.Open("./test/test.cdb")
    if err != nil {
        log.Fatal(err)
    }

    // Create an iterator for the database.
    iter := db.Iter()
    for iter.Next() {
        // Do something with iter.Key()/iter.Value()
    }

    if err := iter.Err(); err != nil {
        log.Fatal(err)
    }
}
```

colinmarc commented:
Yup, looks good!

Commit:
Updated documentation for Iterator

Commit:
Refactored out iterator tests into own file. Added example

JimmyPettersson85 commented:
- Updated docs
- Refactored tests
- Added example for Iterator

colinmarc commented:
Awesome, thank you so so much! Merged in da395d3bba80294e3020c931109a32005a6d842d.

JimmyPettersson85 commented:
Great, thanks!
